src/samVisualize/rendering/navigation.ts

```diff
@@ -0,0 +1,34 @@
+/*!
+ * Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+import * as vscode from 'vscode'
+import * as yaml from 'yaml'
+/**
+ * Maps a resource name to the line number in the template in which it is defined
+ */
+export type ResourceLineMap = {
+    [index: string]: number
+}
+/**
+ * Maps all the resources in a given CFN template to the line numbers at which they are declared
+ * @param textDocument A TextDocument containing a CFN template from which we want to extract a ResourceLineMap
+ * @returns A ResourceLineMap containing all the resources in the given template
+ */
+export function generateResourceLineMap(cfnTemplate: vscode.TextDocument): ResourceLineMap {
+    const lineMap: ResourceLineMap = {}
+    const contents = yaml.parseDocument(cfnTemplate.getText()).contents
+
+    // This check is necessary because yaml.parseDocument could return a Scalar or just null
+    if (contents && 'items' in contents) {
```

Comment:
Will some other logic notify the user that the input is invalid or is it 'silent'? It looks like anything that calls `generateResourceLineMap` will not know whether or not an error has occurred or if the file is just empty.

Reply:
Yeah, there should be better error handling. For example if those `if` statements fail. But as I said, the yaml is validated before this is called. So this is not the site to report errors to the user. I can't really think of a case where a template follows [template anatomy](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/template-anatomy.html) but `generateResourceLineMap` will fail. Those `if` statements shouldn't fail in the case of anatomically correct template, they're only there to satisfy typescript

Reply:
Updated with better documentation, and returning `undefined` if `Resources` are not found in the template